### PR TITLE
[GCP] Minimal permission for service account and instruction for creation

### PR DESCRIPTION
**Pull request by Michaelvll**
~~Blocked by #2100~~

This PR fixes the permission checking and the creating instruction for the service account.

Current behavior:
1. Check skypilot-v1 service account, if permissions satisfied, use it.
2. Check ray-autoscaler-sa service account, if permissions satisfied, use it.
3. If none of above succeeded, try to create the skypilot-v1 service account, and assign full permission roles to it.

TODO:
- [x] backward compatibility issue for the project which only has `ray-autoscaler-sa-` service account and the users do not have the permission to create a new service account.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] Create a service account with the name `skypilot-v1` and test the launching and termination from a account with minimal permission
  - [x] `sky launch -c test-with-only-ray-sa --cloud gcp --cpus 2 -i 0 echo hi` with minimal account
  - [x] `sky launch -c test-with-skypilot-v1-missed-permission --cloud gcp --cpus 2 -i 0 echo hi`  with minimal account (fallback to ray)
  - [x] `sky launch -c test-with-skypilot-ray-missed-permission --cloud gcp --cpus 2 -i 0 echo hi`  with minimal account (should skip GCP)
  - [x] `sky launch -c test-with-skypilot-ray-missed-permission --cloud gcp --cpus 2 -i 0 echo hi`  with full account (should correctly add permission to skypilot-v1)
  - [x] `sky launch -c test-with-skypilot-v1-minimal-perm --cloud gcp --cpus 2 -i 0 echo hi`  with minimal account
  - [x] `sky launch -c test-with-no-sa --cloud gcp --cpus 2 -i 0 echo hi`  with minimal account (should fail)
  - [x] `sky launch -c test-with-no-sa --cloud gcp --cpus 2 -i 0 echo hi`  with full account
- [x] Backward compatibility:
  - [x] `sky launch --cloud gcp -c ray-sa --cpus 2` on master
  - [x] switch to this PR and with skypilot-v1 service account: `sky launch --cloud gcp -c ray-sa --cpus 2`
  - [x] `sky launch --cloud gcp -c ray-sa --cpus 2 -i 0` on master
  - [x] switch to this PR and with skypilot-v1 service account: `sky launch --cloud gcp -c ray-sa --cpus 2` and make sure the cluster is normally autostopped.
- [x] All smoke tests: `pytest tests/test_smoke.py` with the service account

### CODE REVIEW

**Comment by concretevitamin** [comment]
cc @bilelomrani1

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll. A few questions.

File: sky/backends/cloud_vm_ray_backend.py [hunk 1]
```diff
@@ -723,6 +723,16 @@ def _update_blocklist_on_gcp_error(
                 logger.info('Skipping all regions due to disk size issue.')
                 self._blocked_resources.add(
                     launchable_resources.copy(region=None, zone=None))
+            elif ('Policy update access denied.' in httperror_str[0] or
+                  'IAM_PERMISSION_DENIED' in httperror_str[0]):
+                logger.info('Skipping all regions due to service account not '
+                            'having the required permissions and the user '
+                            'account does not have enough permission to '
+                            'update it. Please contact your administrator and '
+                            'check out: https://skypilot.readthedocs.io/en/latest/cloud-setup/cloud-permissions.html#gcp')  # pylint: disable=line-too-long
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:732]: Can we print the detailed underlying error too?

File: docs/source/cloud-setup/cloud-permissions/index.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+Minimal Cloud Permissions
+=================================
+
+.. note::
+
+    The instructions below are completely **optional** and are not needed to get started with SkyPilot. These steps are for users who want to make SkyPilot use non-default, restricted permissions.
+
+SkyPilot is designed to minimize cloud administration and setup steps. It should work out of the box for most users who own the root (admin) account of their clouds, if they follow the steps in :ref:`Installation <installation>` section.
+
+However, in some cases, the user may want to **limit the permissions** of the account used by SkyPilot. This is especially true if the user is using a shared cloud account. The following sections describe the permissions required by SkyPilot for each cloud provider.
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions/index.rst:12]:
```suggestion
However, in some cases, the cloud administrator may want to **limit the permissions** of the account used by SkyPilot. This is especially true if the user is using a shared cloud account. The following sections describe the permissions required by SkyPilot for each cloud provider.
```

File: docs/source/cloud-setup/cloud-permissions/gcp.rst [hunk 3]
```diff
@@ -0,0 +1,152 @@
+.. _cloud-permissions-gcp:
+
+GCP
+=============
+
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions/gcp.rst:5]: Generally, the administrator can choose among three "levels" of permissions, from the most permissive and least setup effort, to the least permissive and more setup effort:

- **Default**: no setup, give users Owner-level permissions (i.e., you do not need to follow the instructions here)
- **Medium-level**: easy setup, with a medium set of permissions
- **Minimal**: more setup, with the minimal set of permissions

(Please check if correct)

File: docs/source/cloud-setup/cloud-permissions/gcp.rst [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  roles/tpu.admin
+
+You can grant those accesses via the GCP's `IAM & Admin console <https://console.cloud.google.com/iam-admin/iam>`__.
+
+
+Minimal Permissions
+-----------------------
+
+The :ref:`Medium Permissions <medium-permissions>` assigns strong admin permissions to the user for some GCP services. It is fine for most users, but if you would like to grant fine-grained minimal permissions to your users in your organization / project, you can create a custom role with the permissions required by following the steps below:
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions/gcp.rst:34]:
```suggestion
The :ref:`Medium Permissions <medium-permissions>` assigns admin permissions for some GCP services to the user.  If you would like to grant finer-grained and more minimal permissions to your users in your organization / project, you can create a custom role by following the steps below:
```

File: docs/source/cloud-setup/cloud-permissions/gcp.rst [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+User
+~~~~~~~~~~~~
+
+1. Go to the GCP's `IAM & Admin console <https://console.cloud.google.com/iam-admin/roles>`__ and click on **Create Role**.
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions/gcp.rst:39]:
```suggestion
1. Go to GCP's `IAM & Admin console <https://console.cloud.google.com/iam-admin/roles>`__ and click on **Create Role**.
```

File: docs/source/cloud-setup/cloud-permissions/gcp.rst [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+    storage.buckets.create
+    storage.buckets.get
+    storage.buckets.delete
+    storage.objects.create
+    storage.objects.delete
+    storage.objects.get
+    storage.objects.list
+
+5. **Optional**: If the user need to access TPU VMs, you can additionally add the following permissions (the following may not be exhaustive, please file an issue if you find any missing permissions):
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions/gcp.rst:96]:
```suggestion
5. **Optional**: If the user needs to access TPU VMs, you can additionally add the following permissions (the following may not be exhaustive, please file an issue if you find any missing permissions):
```

File: docs/source/cloud-setup/cloud-permissions/gcp.rst [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+.. image:: ../../images/screenshots/gcp/create-iam.png
+    :width: 80%
+    :align: center
+    :alt: GCP Grant Access
+
+8. The user should receive an invitation to the project and should be able to setup SkyPilot by following the instructions in :ref:`Installation <installation-gcp>`.
+
+.. note::
+
+    Similar as the AWS, the user created with the above minimal permissions will not be able to create service accounts to be assigned to the instances. The admin need to follow the instruction :ref:`below <gcp-service-account-creation>` to create a service account shared by all the users in the project.
```
Comment (concretevitamin) [inline docs/source/cloud-setup/cloud-permissions/gcp.rst:121]:
```suggestion
    The user created with the above minimal permissions will not be able to create service accounts to be assigned to SkyPilot instances. 
    
    The admin needs to follow the :ref:`instruction below <gcp-service-account-creation>` to create a service account to be shared by all users in the project.
```

File: sky/skylet/providers/gcp/config.py [hunk 8]
```diff
@@ -333,6 +344,79 @@ def _configure_project(config, crm):
...
+    logger.info(f"_configure_iam_role: Checking permissions for {email}...")
+
+    # Check the roles first, as checking the permission requires more API calls and
+    # permissions.
+    for role in required_roles:
+        role_exists = False
+        for binding in policy["bindings"]:
+            if binding["role"] == role:
+                if member_id not in binding["members"]:
+                    print(f"Role missing: {role}")
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:373]: remnant?

File: sky/skylet/providers/gcp/config.py [hunk 9]
```diff
@@ -333,6 +344,79 @@ def _configure_project(config, crm):
...
+                    raise
+            else:
+                permissions = role_definition["includedPermissions"]
+            required_permissions -= set(permissions)
+        if not required_permissions:
+            break
+    logger.info(f"_configure_iam_role: missing permisisons {required_permissions}")
+    if not required_permissions:
+        # All required permissions are already granted. We don't need to check
+        # the roles below.
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:414]: Second sentence seems outdated

File: sky/skylet/providers/gcp/config.py [hunk 10]
```diff
@@ -333,6 +344,79 @@ def _configure_project(config, crm):
...
+                    )
+                    permissions = []
+                else:
+                    raise
+            else:
+                permissions = role_definition["includedPermissions"]
+            required_permissions -= set(permissions)
+        if not required_permissions:
+            break
+    logger.info(f"_configure_iam_role: missing permisisons {required_permissions}")
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:411]: Move to L416?

File: sky/skylet/providers/gcp/config.py [hunk 11]
```diff
@@ -345,30 +429,69 @@ def _configure_iam_role(config, crm, iam):
...
-        )
+        if ray_satisfied:
+            service_account = ray_service_account
+            satisfied = ray_satisfied
+        elif service_account is None:
+            logger.info(
+                "_configure_iam_role: "
+                "Creating new service account {}".format(SKYPILOT_SERVICE_ACCOUNT_ID)
+            )
+            service_account = _create_service_account(
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:478]: Is it correct to say that, for a new user without any SA and with no permission to create SAs, they will hit here and fail? May want a comment.

File: sky/skylet/providers/gcp/config.py [hunk 12]
```diff
@@ -839,7 +962,7 @@ def _get_service_account(account, config, iam):
     try:
         service_account = iam.projects().serviceAccounts().get(name=full_name).execute()
     except errors.HttpError as e:
-        if e.resp.status != 404:
+        if e.resp.status not in [403, 404]:
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:965]: What's 403?

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll. Some questions.

File: sky/backends/cloud_vm_ray_backend.py (hunk 1 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py]: Good point! Fixed. Thanks!

File: sky/skylet/providers/gcp/config.py
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py:1]: Yes. Ran the following tests and it seems fine:
1. `sky launch -c min --cloud gcp` on master
2. `sky launch -c min` on this branch

Also, 
1. `sky launch -c min --cloud gcp` on master
2. `sky stop min`
3. `sky launch -c min` on this branch

File: sky/skylet/providers/gcp/config.py (hunk 8 above)
Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py]: Oops. Good catch!

File: sky/skylet/providers/gcp/config.py (hunk 11 above)
Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py:477]: Yes, good point! Added. Thanks!

File: sky/skylet/providers/gcp/config.py (hunk 10 above)
Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py]: Done. Thanks!

File: sky/skylet/providers/gcp/config.py (hunk 9 above)
Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py]: Ahh, good catch! Thanks!

File: docs/source/cloud-setup/cloud-permissions/aws.rst [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    :alt: AWS Add User
+
+2. In the **Permissions options** section, select "Attach existing policies directly"; Click on the **Create Policy**. This opens another window to create an IAM policy.
+
+.. image:: ../../images/screenshots/aws/aws-create-policy.png
+    :width: 80%
+    :align: center
+    :alt: AWS Create Policy
+
+3. Choose "JSON" tab and place the following policy into the box. Replace the ``<account-ID-without-hyphens>`` with your AWS account ID. You can find your AWS account ID by clicking on the upper right corner of the console.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply docs/source/cloud-setup/cloud-permissions/aws.rst]: Done. Thanks!

**Review by concretevitamin — APPROVED** [review]
Thanks a bunch for adding this @Michaelvll !

**Pull request merged**